cluster/centos/config-default.sh

```diff
@@ -14,24 +14,74 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+readonly root=$(dirname "${BASH_SOURCE}")
+
 ## Contains configuration values for the CentOS cluster
 # The user should have sudo privilege
 export MASTER=${MASTER:-"centos@172.10.0.11"}
 export MASTER_IP=${MASTER#*@}
 
+# Define all your master nodes,
+# And separated with blank space like <user_1@ip_1> <user_2@ip_2> <user_3@ip_3>.
+# The user should have sudo privilege
+export MASTERS="${MASTERS:-$MASTER}"
+
+# length-of <arg0>
+# Get the length of specific arg0, could be a space-separate string or array.
+function length-of() {
+  local len=0
+  for part in $1; do
+    let len++
+  done
+  echo $len
+}
+# Number of nodes in your cluster.
+export NUM_MASTERS="${NUM_MASTERS:-$(length-of "$MASTERS")}"
+
+# Define advertise address of masters, could be a load balancer address.
+# If not provided, the default is ip of first master node.
+function default-advertise-address() {
+  # get the first master node
+  local masters_array=(${MASTERS})
+  local master=${masters_array[0]}
+  echo ${master#*@}
+}
+export MASTER_ADVERTISE_ADDRESS="${MASTER_ADVERTISE_ADDRESS:-$(default-advertise-address)}"
+export MASTER_ADVERTISE_IP="${MASTER_ADVERTISE_IP:-$(getent hosts "${MASTER_ADVERTISE_ADDRESS}" | awk '{print $1; exit}')}"
+
 # Define all your minion nodes,
 # And separated with blank space like <user_1@ip_1> <user_2@ip_2> <user_3@ip_3>.
 # The user should have sudo privilege
-export NODES=${NODES:-"centos@172.10.0.12 centos@172.10.0.13"}
+export NODES="${NODES:-"centos@172.10.0.12 centos@172.10.0.13"}"
 
 # Number of nodes in your cluster.
-export NUM_NODES=${NUM_NODES:-2}
+export NUM_NODES="${NUM_NODES:-$(length-of "$NODES")}"
 
 # Should be removed when NUM_NODES is deprecated in validate-cluster.sh
-export NUM_NODES=${NUM_NODES}
+export NUM_NODES="${NUM_NODES}"
 
 # By default, the cluster will use the etcd installed on master.
-export ETCD_SERVERS=${ETCD_SERVERS:-"http://$MASTER_IP:2379"}
+function concat-etcd-servers() {
+  local etcd_servers=""
+  for master in ${MASTERS}; do
+    local master_ip=${master#*@}
+    local prefix=""
+    if [ -n "$etcd_servers" ]; then
+      prefix="${etcd_servers},"
+    fi
+    etcd_servers="${prefix}http://$master_ip:2379"
+  done
+
+  echo $etcd_servers
+}
+export ETCD_SERVERS="$(concat-etcd-servers)"
+
+# By default, etcd cluster will use etcd-discovery
+#   https://github.com/coreos/etcd/blob/master/Documentation/op-guide/clustering.md#etcd-discovery
+export NUM_ETCD_NODES="${NUM_MASTERS}"
+export ETCD_DISCOVERY="${ETCD_DISCOVERY:-$(curl -sSL https://discovery.etcd.io/new?size=${NUM_ETCD_NODES})}"
```

Comment:
So, you are using etcd discovery to create etcd cluster. In GCE, we are using [runtime reconfiguration](https://coreos.com/etcd/docs/latest/v2/runtime-configuration.html). Your approach creates a dependency on external etcd discovery service, which, in my opinion, is better to prevent.